// @chakra-ui/color-mode/dist/esm/color-mode.utils.js
import { isBrowser, noop } from "@chakra-ui/utils";
var classNames = {
  light: "chakra-ui-light",
  dark: "chakra-ui-dark"
};

/**
 * SSR: Graceful fallback for the `body` element
 */
var mockBody = {
  classList: {
    add: noop,
    remove: noop
  }
};
export var body = isBrowser ? document.body : mockBody;
/**
 * Function to add/remove class from `body` based on color mode
 */

export function syncBodyClassName(isDark) {
  body.classList.add(isDark ? classNames.dark : classNames.light);
  body.classList.remove(isDark ? classNames.light : classNames.dark);
}
/**
 * Check if JS media query matches the query string passed
 */

function getMediaQuery(query) {
  var mediaQueryList = window.matchMedia == null ? void 0 : window.matchMedia(query);
  var matches = !!mediaQueryList.media === mediaQueryList.matches;
  return matches;
}

export var queries = {
  light: "(prefers-color-scheme: light)",
  dark: "(prefers-color-scheme: dark)"
};
export var lightQuery = queries.light;
export var darkQuery = queries.dark;
export function getColorScheme() {
  var isDark = getMediaQuery(queries.dark);
  return isDark ? "dark" : "light";
}
/**
 * Adds system os color mode listener, and run the callback
 * once preference changes
 */

export function addListener(fn) {
  if (!("matchMedia" in window)) {
    return noop;
  }

  var mediaQueryList = window.matchMedia(queries.dark);

  var listener = () => {
    fn(!!mediaQueryList.matches ? "dark" : "light");
  };

  listener();
  mediaQueryList.addListener(listener);
  return () => {
    mediaQueryList.removeListener(listener);
  };
}
