// @chakra-ui/form-control/dist/esm/use-form-control.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import { useFormControlContext } from "./form-control";
import { ariaAttr, dataAttr, callAllHandlers, omit } from "@chakra-ui/utils";

/**
 * React hook that provides the props that should be spread on to
 * input fields (`input`, `select`, `textarea`, etc.).
 *
 * It provides a convenient way to control a form fields, validation
 * and helper text.
 */
export function useFormControl(props) {
  var _props$id;

  var field = useFormControlContext();
  var describedBy = [];

  if (field == null ? void 0 : field.isInvalid) {
    /**
     * Error message must be described first
     * in all scenarios
     */
    if (describedBy.length > 0) {
      describedBy.unshift(field.feedbackId);
    } else {
      describedBy.push(field.feedbackId);
    }
  }

  if (field == null ? void 0 : field.hasHelpText) describedBy.push(field.helpTextId);
  var ariaDescribedBy = describedBy.join(" ");
  var cleanProps = omit(props, ["isInvalid", "isDisabled", "isReadOnly", "isRequired"]);
  return _extends({}, cleanProps, {
    id: (_props$id = props.id) != null ? _props$id : field == null ? void 0 : field.id,
    disabled: props.disabled || props.isDisabled || (field == null ? void 0 : field.isDisabled),
    readOnly: props.readOnly || props.isReadOnly || (field == null ? void 0 : field.isReadOnly),
    required: props.required || props.isRequired || (field == null ? void 0 : field.isRequired),
    "aria-invalid": ariaAttr(props.isInvalid || (field == null ? void 0 : field.isInvalid)),
    "aria-required": ariaAttr(props.isRequired || (field == null ? void 0 : field.isRequired)),
    "aria-readonly": ariaAttr(props.isReadOnly || (field == null ? void 0 : field.isReadOnly)),
    "aria-describedby": ariaDescribedBy || undefined,
    onFocus: callAllHandlers(field == null ? void 0 : field.onFocus, props.onFocus),
    onBlur: callAllHandlers(field == null ? void 0 : field.onBlur, props.onBlur)
  });
}
export function useFormControlLabel(props) {
  var _props$id2, _props$htmlFor;

  var field = useFormControlContext();
  return _extends({}, props, {
    "data-focus": dataAttr(field == null ? void 0 : field.isFocused),
    "data-disabled": dataAttr(field == null ? void 0 : field.isDisabled),
    "data-invalid": dataAttr(field == null ? void 0 : field.isInvalid),
    "data-loading": dataAttr(field == null ? void 0 : field.isLoading),
    "data-readonly": dataAttr(field == null ? void 0 : field.isReadOnly),
    id: (_props$id2 = props.id) != null ? _props$id2 : field == null ? void 0 : field.labelId,
    htmlFor: (_props$htmlFor = props.htmlFor) != null ? _props$htmlFor : field == null ? void 0 : field.id
  });
}
