// @chakra-ui/color-mode/dist/cjs/color-mode.utils.js
"use strict";

exports.__esModule = true;
exports.syncBodyClassName = syncBodyClassName;
exports.getColorScheme = getColorScheme;
exports.addListener = addListener;
exports.darkQuery = exports.lightQuery = exports.queries = exports.body = void 0;

var _utils = require("@chakra-ui/utils");

var classNames = {
  light: "chakra-ui-light",
  dark: "chakra-ui-dark"
};

/**
 * SSR: Graceful fallback for the `body` element
 */
var mockBody = {
  classList: {
    add: _utils.noop,
    remove: _utils.noop
  }
};
var body = _utils.isBrowser ? document.body : mockBody;
/**
 * Function to add/remove class from `body` based on color mode
 */

exports.body = body;

function syncBodyClassName(isDark) {
  body.classList.add(isDark ? classNames.dark : classNames.light);
  body.classList.remove(isDark ? classNames.light : classNames.dark);
}
/**
 * Check if JS media query matches the query string passed
 */


function getMediaQuery(query) {
  var mediaQueryList = window.matchMedia == null ? void 0 : window.matchMedia(query);
  var matches = !!mediaQueryList.media === mediaQueryList.matches;
  return matches;
}

var queries = {
  light: "(prefers-color-scheme: light)",
  dark: "(prefers-color-scheme: dark)"
};
exports.queries = queries;
var lightQuery = queries.light;
exports.lightQuery = lightQuery;
var darkQuery = queries.dark;
exports.darkQuery = darkQuery;

function getColorScheme() {
  var isDark = getMediaQuery(queries.dark);
  return isDark ? "dark" : "light";
}
/**
 * Adds system os color mode listener, and run the callback
 * once preference changes
 */


function addListener(fn) {
  if (!("matchMedia" in window)) {
    return _utils.noop;
  }

  var mediaQueryList = window.matchMedia(queries.dark);

  var listener = function listener() {
    fn(!!mediaQueryList.matches ? "dark" : "light");
  };

  listener();
  mediaQueryList.addListener(listener);
  return function () {
    mediaQueryList.removeListener(listener);
  };
}
