// @chakra-ui/transition/dist/esm/scale-fade.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

import * as React from "react";
import { __DEV__ } from "@chakra-ui/utils";
import { Transition } from "./transition";

function getTransitionStyles(initialScale) {
  return {
    init: {
      opacity: 0,
      transform: "scale(" + initialScale + ")"
    },
    entered: {
      opacity: 1,
      transform: "scale(1)"
    },
    exiting: {
      opacity: 0,
      transform: "scale(" + initialScale + ")"
    }
  };
}

export var ScaleFade = props => {
  var {
    initialScale = 0.9,
    timeout = 150
  } = props,
      rest = _objectWithoutPropertiesLoose(props, ["initialScale", "timeout"]);

  var styles = getTransitionStyles(initialScale);
  return /*#__PURE__*/React.createElement(Transition, _extends({
    styles: styles,
    transition: "all " + timeout + "ms cubic-bezier(0.45, 0, 0.40, 1)",
    timeout: {
      enter: 0,
      exit: timeout
    },
    unmountOnExit: true
  }, rest));
};

if (__DEV__) {
  ScaleFade.displayName = "ScaleFade";
}
