// @chakra-ui/transition/dist/esm/slide-fade.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

import * as React from "react";
import { __DEV__ } from "@chakra-ui/utils";
import { Transition } from "./transition";

function getTransitionStyles(initialOffset) {
  return {
    init: {
      opacity: 0,
      transform: "translateY(" + initialOffset + ")"
    },
    entered: {
      opacity: 1,
      transform: "translateY(0px)"
    },
    exiting: {
      opacity: 0,
      transform: "translateY(" + initialOffset + ")"
    }
  };
}

export var SlideFade = props => {
  var {
    initialOffset = "20px",
    timeout = 150
  } = props,
      rest = _objectWithoutPropertiesLoose(props, ["initialOffset", "timeout"]);

  var styles = getTransitionStyles(initialOffset);
  return /*#__PURE__*/React.createElement(Transition, _extends({
    styles: styles,
    transition: "all " + timeout + "ms cubic-bezier(0.4, 0.14, 0.3, 1)",
    timeout: {
      enter: 0,
      exit: timeout
    }
  }, rest));
};

if (__DEV__) {
  SlideFade.displayName = "SlideFade";
}
