// @chakra-ui/layout/dist/esm/wrap.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

import { chakra, css, forwardRef, useTheme } from "@chakra-ui/system";
import { getValidChildren, mapResponsive, __DEV__ } from "@chakra-ui/utils";
import * as React from "react";

/**
 * Layout component used to stack elements that differ in length
 * and are liable to wrap.
 *
 * Common use cases:
 * - Buttons that appear together at the end of forms
 * - Lists of tags and chips
 *
 * @see Docs https://chakra-ui.com/components/wrap
 */
export var Wrap = /*#__PURE__*/forwardRef(function Wrap(props, ref) {
  var {
    spacing = "0.5rem",
    children,
    justify,
    direction,
    align
  } = props,
      rest = _objectWithoutPropertiesLoose(props, ["spacing", "children", "justify", "direction", "align"]);

  var theme = useTheme();
  var itemSpacing = mapResponsive(spacing, value => {
    var {
      margin
    } = css({
      margin: value
    })(theme);
    return "calc(" + margin + " / 2)";
  });
  var groupSpacing = mapResponsive(spacing, value => {
    var {
      margin
    } = css({
      margin: value
    })(theme);
    return "calc(" + margin + " / 2 * -1)";
  });
  var validChildren = getValidChildren(children);
  var clones = validChildren.map((child, index) => /*#__PURE__*/React.createElement(chakra.li, {
    key: index,
    margin: itemSpacing
  }, child));
  var styles = {
    display: "flex",
    flexWrap: "wrap",
    justifyContent: justify,
    alignItems: align,
    flexDirection: direction,
    listStyleType: "none",
    padding: "0",
    margin: groupSpacing
  };
  return /*#__PURE__*/React.createElement(chakra.div, _extends({
    ref: ref
  }, rest), /*#__PURE__*/React.createElement(chakra.ul, {
    __css: styles
  }, clones));
});

if (__DEV__) {
  Wrap.displayName = "Wrap";
}
